### PR TITLE
can set a patchData w. an all_primal centering

### PR DESCRIPTION
**Pull request by rochSmets**
This is just the beginning of the work... I will continue once we will be agree with the way to do it.
My notebook to test it is here : 
https://github.com/rochSmets/farJunk/blob/main/nb/ohm.ipynb

## Summary by CodeRabbit

- **New Features**
  - Introduced advanced mathematical operations for vector and scalar fields, including dot product, cross product, square root, gradient, and modulus calculations.
  - Added new methods and functions for handling hierarchical data structures, including renaming, arithmetic operations, and domain slicing.

- **Enhancements**
  - Updated `PatchHierarchy` class with new methods for mapping quantities to spatial directions and retrieving patch data names.
  - Improved field computation methods to handle additional parameters and new arithmetic operations.

### CODE REVIEW

File: pyphare/pyphare/pharesee/run.py [hunk 1]
```diff
@@ -131,6 +132,185 @@ def _compute_divB(patchdatas, **kwargs):
...
+
+    ds_shape = list(ds.shape)
+    for i in range(ndim):
+        if patchdata.centerings[i] == 'dual':
+            ds_shape[i] += 1
+
+    ds_all_primal = np.full(ds_shape, np.nan)
+
+    if ndim == 2:
+         ds_all_primal[nb_ghosts : -nb_ghosts, nb_ghosts : -nb_ghosts] = \
```
Comment (PhilipDeegan) [inline pyphare/pyphare/pharesee/run.py:148]: maybe something like [this](https://gist.github.com/PhilipDeegan/a7f423955d43c577b20262c0c58c9b11) could make these selections a bit "DRY"-er

variable names not set in stone

File: pyphare/pyphare/pharesee/hierarchy.py [hunk 2]
```diff
@@ -1070,6 +1085,172 @@ def dist_plot(self, **kwargs):
...
+        return "zob"
+
+
+def _compute_add(patch_datas, **kwargs):
+
+    ref_name = next(iter(patch_datas.keys()))
+
+    dset_x = patch_datas["self_x"].dataset[:] * patch_datas["operand_x"].dataset[:]\
+    dset_y = patch_datas["self_y"].dataset[:] * patch_datas["operand_y"].dataset[:]\
+    dset_z = patch_datas["self_z"].dataset[:] * patch_datas["operand_z"].dataset[:]\
```
Comment (PhilipDeegan) [inline pyphare/pyphare/pharesee/hierarchy.py:1229]: trailing slashes are causing errors here

File: pyphare/pyphare/core/operators.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+def _compute_dot_product(patch_datas, **kwargs):
+
+    ref_name = next(iter(patch_datas.keys()))
+
+    dset = patch_datas["left_x"].dataset[:] * patch_datas["right_x"].dataset[:]\
+         + patch_datas["left_y"].dataset[:] * patch_datas["right_y"].dataset[:]\
+         + patch_datas["left_z"].dataset[:] * patch_datas["right_z"].dataset[:]
+
+    return ({"name": 'scalar', "data": dset, "centering": patch_datas[ref_name].centerings},)
```
Comment (PhilipDeegan) [inline pyphare/pyphare/core/operators.py:16]: this (line 16) looks like it returns a tuple of size one, is that the case? and if so, is that required?

File: pyphare/pyphare/pharesee/run.py [hunk 4]
```diff
@@ -2,7 +2,8 @@
 
 import numpy as np
 from pyphare.pharesee.hierarchy import compute_hier_from
-from pyphare.pharesee.hierarchy import compute_hier_from_
+# from pyphare.pharesee.hierarchy import compute_hier_from_
+from pyphare.pharesee.hierarchy import ScalarField, VectorField
```
[the remark this answers is not in the document]
Comment (PhilipDeegan) [reply pyphare/pyphare/pharesee/run.py:6]: @rochSmets I think this scanning note is incorrect, but it does show an inconsistency we can probably fix

the final form should probably be like

```
from pyphare.pharesee.hierarchy import (
    compute_hier_from,
    flat_finest_field,
    hierarchy_from,
    ScalarField,
    VectorField,
)
```

the mix of normal and relatives imports seems to be confusing it

File: pyphare/pyphare/pharesee/run.py [hunk 5]
```diff
@@ -371,10 +573,26 @@ def GetPi(self, time, merged=False, interp="nearest"):
         Pi = compute_hier_from(_compute_pressure, (M, massDensity, Vi))
         return self._get(Pi, time, merged, interp)
 
-    def GetJ(self, time, merged=False, interp="nearest"):
+    def GetPe(self, time, merged=False, interp="nearest", all_primal=False):
+        # TODO Te should come from the run... as an attribute of a hierarchy ?
+        Te = 1.0
+        hier = self._get_hierarchy(time, "ions_density.h5")
```
Comment (PhilipDeegan) [inline pyphare/pyphare/pharesee/run.py:579]: with the latest PR merged putting the simulation in diagnostic files you should be able to change this to

```
        hier = self._get_hierarchy(time, "ions_density.h5")
        Te = hier.sim.electrons.closure.Te
```

File: pyphare/pyphare/core/operators.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+    else:
+        raise RuntimeError("type of hierarchy not yet considered")
+
+    hl = rename(hier_left, names_left)
+    hr = rename(hier_right, names_right)
+
+    h = compute_hier_from(_compute_dot_product, (hl, hr),)
+
+    hier_left = rename(hl, names_left_kept)
```
[the remark this answers is not in the document]
Comment (rochSmets) [reply pyphare/pyphare/core/operators.py:92]: Yes but this rename is needed to preserve the original names in the hierarchy (modified for the purpose of the operation on it)

File: pyphare/pyphare/pharesee/run.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rochSmets) [reply pyphare/pyphare/pharesee/run.py:6]: done,

File: pyphare/pyphare/pharesee/hierarchy.py [hunk 7]
```diff
@@ -1069,6 +1084,209 @@
...
+class ScalarField(PatchHierarchy):
+
+    name = 'scalar'
+
+    def __init__(self, patch_levels, domain_box, **kwargs):
+        refinement_ratio = kwargs.get("refinement_ratio", 2)
+        time = kwargs.get("time", .0)
+        data_files = kwargs.get("data_files", None)
+
+        super().__init__(patch_levels, domain_box, refinement_ratio, time, data_files)
```
[the remark this answers is not in the document]
Comment (rochSmets) [reply pyphare/pyphare/pharesee/hierarchy.py:1126]: static name removed

File: pyphare/pyphare/core/operators.py (hunk 3 above)
Comment (rochSmets) [reply pyphare/pyphare/core/operators.py:16]: Yes, if I'm correct, the `compute` call back returns a set of dict containing the `name`, `data` and `centering`. At the end of the game, this is `new_patchdatas_from` which call the `compute callback, and put the results in `datas`. As we iterate in the elements of `datas`, I think it has to be a tuple.

File: pyphare/pyphare/pharesee/run.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rochSmets) [reply pyphare/pyphare/pharesee/run.py:6]: done

File: pyphare/pyphare/pharesee/run.py [hunk 8]
```diff
@@ -130,6 +131,185 @@
...
+    ds_all_primal = np.full(ds_shape, np.nan)
+
+    if ndim == 2:
+        ds_all_primal[nb_ghosts:-nb_ghosts, nb_ghosts:-nb_ghosts] = \
+            0.5*(np.asarray(ds[nb_ghosts:-nb_ghosts, nb_ghosts-1:-nb_ghosts]) +
+                 np.asarray(ds[nb_ghosts:-nb_ghosts, nb_ghosts:-nb_ghosts+1]))
+    else:
+        raise RuntimeError("dimension not yet implemented")
+
+    return ds_all_primal
```
[the remark this answers is not in the document]
Comment (rochSmets) [reply pyphare/pyphare/pharesee/run.py:153]: yep, see the first comment of @PhilipDeegan

File: pyphare/pyphare/pharesee/run.py [hunk 9]
```diff
@@ -130,6 +131,185 @@
...
+        elif pd_name == "Jy":  # TODO same centering as Ey... only for Yee !
+            dataset = _compute_Ey_to_primal(pd, nb_ghosts=nb_ghosts)
+        elif pd_name == "Jz":  # TODO same centering as Ez... only for Yee !
+            dataset = _compute_Ez_to_primal(pd, nb_ghosts=nb_ghosts)
+        else:  # TODO need to figure out what to do for pressure of each specie
+            raise RuntimeError("patchdata name unknown")
+        pd_attrs.append({"name": name, "data": dataset, "centering": centerings})
+        # break
+
+    return tuple(pd_attrs)
```
[the remark this answers is not in the document]
Comment (rochSmets) [reply pyphare/pyphare/pharesee/run.py:310]: yes... can we manage the definition of the grid (yee or not) in a nicer way ?

File: pyphare/pyphare/core/operators.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rochSmets) [reply pyphare/pyphare/core/operators.py:16]: already done in the `dot` method

File: pyphare/pyphare/pharesee/hierarchy.py [hunk 10]
```diff
@@ -563,6 +563,18 @@
...
+        if qty in ['rho', 'tags']:
+            return 'scalar'
+        elif qty in ['Bx', 'Ex', 'Fx', 'Vx']:
+            return 'x'
+        elif qty in ['By', 'Ey', 'Fy', 'Vy']:
+            return 'y'
+        elif qty in ['Bz', 'Ez', 'Fz', 'Vz']:
+            return 'z'
+        else:
+            raise ValueError("{} is not a valid quantity".format(qty))
```
[the remark this answers is not in the document]
Comment (rochSmets) [reply pyphare/pyphare/pharesee/hierarchy.py:576]: done with a `dict`... I dont know how to do this w. a `map`

File: pyphare/pyphare/core/operators.py [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+    hier_right = rename(hr, names_right_kept)
+
+    return ScalarField(h.patch_levels, h.domain_box,
+                       refinement_ratio=h.refinement_ratio,
+                       time=h.times()[0],
+                       data_files=h.data_files)
```
[the remark this answers is not in the document]
Comment (rochSmets) [reply pyphare/pyphare/core/operators.py:98]: they are for name preserving...

File: pyphare/pyphare/pharesee/run.py [hunk 12]
```diff
@@ -370,10 +571,26 @@
         Pi = compute_hier_from(_compute_pressure, (M, massDensity, Vi))
         return self._get(Pi, time, merged, interp)
 
-    def GetJ(self, time, merged=False, interp="nearest"):
+    def GetPe(self, time, merged=False, interp="nearest", all_primal=False):
+        # TODO Te should come from the run... as an attribute of a hierarchy ?
+        Te = 1.0
+        hier = self._get_hierarchy(time, "ions_density.h5")
```
[the remark this answers is not in the document]
Comment (rochSmets) [reply pyphare/pyphare/pharesee/run.py:577]: done

File: pyphare/pyphare/core/operators.py [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+
+    h = compute_hier_from(_compute_cross_product, (hl, hr),)
+
+    hier_left = rename(hl, names_left_kept)
+    hier_right = rename(hr, names_right_kept)
+
+    return VectorField(h.patch_levels, h.domain_box,
+                       refinement_ratio=h.refinement_ratio,
+                       time=h.times()[0],
+                       data_files=h.data_files)
```
[the remark this answers is not in the document]
Comment (rochSmets) [reply pyphare/pyphare/core/operators.py:123]: idem

File: pyphare/pyphare/core/operators.py [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+
+    return sqrt(dot(hier, hier))
+
+
+def grad(hier, **kwargs):
+    assert isinstance(hier, ScalarField)
+    nb_ghosts = list(hier.level(0).patches[0].patch_datas.values())[0].ghosts_nbr[0]
+    h = compute_hier_from(_compute_grad, hier, nb_ghosts=nb_ghosts)
+
+    # TODO the plot of a grad displays only 1 patch if vmin and vmax are not specified... any idea why ?
```
[the remark this answers is not in the document]
Comment (rochSmets) [reply pyphare/pyphare/core/operators.py:146]: fixed

File: pyphare/pyphare/pharesee/run.py (hunk 1 above)
Comment (rochSmets) [reply pyphare/pyphare/pharesee/run.py]: yes, you're right. I have rewritten the '_compute_to_primal' in that spirit

**Comment by rochSmets** [comment]
I still have 2 commits to keep memory of what I did if it happens I need to restore anything. But if this is fine for you, I will squash in a single commit

**Comment by PhilipDeegan** [comment]
> I still have 2 commits to keep memory of what I did if it happens I need to restore anything. But if this is fine for you, I will squash in a single commit

it can all be squashed on merge so you don't need to bother

File: pyphare/pyphare/core/operators.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (PhilipDeegan) [reply pyphare/pyphare/core/operators.py:92]: I think the issue here, which I'm not sure is strictly an issue, is that in python, modifying input parameters may not alter the variable at the call site

image

so you might need to do something like 

```
def dot(hier_left, hier_right, **kwargs):
    ...
    return hier_left, hier_right, ScalarField(h.patch_levels, h.domain_box,
                       refinement_ratio=h.refinement_ratio,
                       time=h.times()[0],
                       data_files=h.data_files)

# somewhere using dot
     hier_left, hier_right, field = dot(...)
```

File: pyphare/pyphare/pharesee/hierarchy.py [hunk 15]
```diff
@@ -1088,6 +1125,244 @@ def dist_plot(self, **kwargs):
...
+            names_self = ["self_x", "self_y", "self_z"]
+            names_other = ["other_x", "other_y", "other_z"]
+        else:
+            raise RuntimeError("type of hierarchy not yet considered")
+
+        h_self = rename(self, names_self)
+        h_other = rename(other, names_other)
+
+        h = compute_hier_from(
+            _compute_sub,
```
Comment (nicolasaunai) [inline pyphare/pyphare/pharesee/hierarchy.py:1238]: is it possible/better to do sub as  add other*-1 ?

File: pyphare/pyphare/core/operators.py [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+
+    hl = rename(hier_left, names_left)
+    hr = rename(hier_right, names_right)
+
+    h = compute_hier_from(
+        _compute_dot_product,
+        (hl, hr),
+    )
+
+    # hier_left = rename(hl, names_left_kept)
```
Comment (nicolasaunai) [inline pyphare/pyphare/core/operators.py:109]: rm both lines?

File: pyphare/pyphare/core/operators.py [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+
+    hl = rename(hier_left, names_left)
+    hr = rename(hier_right, names_right)
+
+    h = compute_hier_from(
+        _compute_cross_product,
+        (hl, hr),
+    )
+
+    # hier_left = rename(hl, names_left_kept)
```
Comment (nicolasaunai) [inline pyphare/pyphare/core/operators.py:140]: rm both lines?

File: pyphare/pyphare/core/operators.py [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+        refinement_ratio=h.refinement_ratio,
+        time=h.times()[0],
+        data_files=h.data_files,
+    )
+
+
+def modulus(hier):
+    assert isinstance(hier, VectorField)
+
+    return sqrt(dot(hier, hier))
```
Comment (nicolasaunai) [inline pyphare/pyphare/core/operators.py:170]: c'est beau

File: pyphare/pyphare/core/operators.py [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+
+
+def grad(hier, **kwargs):
+    assert isinstance(hier, ScalarField)
+    nb_ghosts = list(hier.level(0).patches[0].patch_datas.values())[0].ghosts_nbr[0]
+    h = compute_hier_from(_compute_grad, hier, nb_ghosts=nb_ghosts)
+
+    # TODO the plot of a grad displays only 1 patch if vmin and vmax are not specified... any idea why ?
```
Comment (nicolasaunai) [inline pyphare/pyphare/core/operators.py:178]: I see you told rabbit this is fixed, should we thus rm the comment?

File: pyphare/pyphare/pharesee/run.py [hunk 20]
```diff
@@ -1,10 +1,19 @@
 import os
-
 import numpy as np
-from pyphare.pharesee.hierarchy import compute_hier_from
-from pyphare.pharesee.hierarchy import compute_hier_from_
 
-from .hierarchy import flat_finest_field, hierarchy_from
+import pyphare.pharein as ph
+
+# from pyphare.pharesee.hierarchy import compute_hier_from
```
Comment (nicolasaunai) [inline pyphare/pyphare/pharesee/run.py:6]: rm 3 lines?

Comment (rochSmets) [reply pyphare/pyphare/pharesee/run.py:6]: done

File: pyphare/pyphare/pharesee/run.py (hunk 5 above)
Comment (rochSmets) [reply pyphare/pyphare/pharesee/run.py:647]: done

File: pyphare/pyphare/pharesee/run.py [hunk 21]
```diff
@@ -323,21 +506,42 @@
...
+            return self._get(hier, time, merged, interp)
+        else:
+            h = compute_hier_from(_compute_to_primal, hier, scalar="rho")
+            return ScalarField(h.patch_levels, h.domain_box,
+                               refinement_ratio=h.refinement_ratio,
+                               time=time,
+                               data_files=h.data_files)
 
     def GetN(self, time, pop_name, merged=False, interp="nearest"):
         hier = self._get_hierarchy(time, f"ions_pop_{pop_name}_density.h5")
```
[the remark this answers is not in the document]
Comment (rochSmets) [reply pyphare/pyphare/pharesee/run.py]: `all_primal` appears wherever it is needed.

File: pyphare/pyphare/pharesee/run.py [hunk 22]
```diff
@@ -130,6 +139,178 @@ def _compute_divB(patchdatas, **kwargs):
...
+
+        ds_shape = list(ds.shape)
+        for i in range(ndim):
+            if pd.centerings[i] == "dual":
+                ds_shape[i] += 1
+
+        ds_all_primal = np.full(ds_shape, np.nan)
+        ds_ = np.zeros(ds_shape)
+
+        if pd_name in ["Fx", "Fy", "Fz", "Vx", "Vy", "Vz", "rho", "tags"]:
```
[the remark this answers is not in the document]
Comment (rochSmets) [reply pyphare/pyphare/pharesee/run.py]: the case of all primal quantities is now included in the `_ppp_to_ppp_domain_slicing` function

File: pyphare/pyphare/pharesee/run.py [hunk 23]
```diff
@@ -130,6 +139,178 @@ def _compute_divB(patchdatas, **kwargs):
...
+    """
+    return the slicing for (dual,primal,primal) to (primal,primal,primal)
+    centering that is the centering of Bx on a Yee grid
+    """
+
+    inner, inner_shift_left, inner_shift_right = _inner_slices(nb_ghosts)
+
+    if ndim == 2:
+        inner_all = tuple([inner] * 2)
+        return inner_all, ((inner, inner_shift_left), (inner, inner_shift_right))
```
[the remark this answers is not in the document]
Comment (rochSmets) [reply pyphare/pyphare/pharesee/run.py]: yes ; will do it very soon... before merging !

File: pyphare/pyphare/pharesee/run.py [hunk 24]
```diff
@@ -130,6 +139,178 @@ def _compute_divB(patchdatas, **kwargs):
...
+                ds_shape[i] += 1
+
+        ds_all_primal = np.full(ds_shape, np.nan)
+        ds_ = np.zeros(ds_shape)
+
+        if pd_name in ["Fx", "Fy", "Fz", "Vx", "Vy", "Vz", "rho", "tags"]:
+            ds_all_primal = np.asarray(patchdatas[pd_name].dataset)
+        elif pd_name == "Bx":
+            inner, chunks = _pdd_to_ppp_domain_slicing(nb_ghosts, ndim)
+            for chunk in chunks:
```
[the remark this answers is not in the document]
Comment (rochSmets) [reply pyphare/pyphare/pharesee/run.py]: done

File: pyphare/pyphare/pharesee/run.py [hunk 25]
```diff
@@ -130,6 +139,178 @@ def _compute_divB(patchdatas, **kwargs):
...
+        for i in range(ndim):
+            if pd.centerings[i] == "dual":
+                ds_shape[i] += 1
+
+        ds_all_primal = np.full(ds_shape, np.nan)
+        ds_ = np.zeros(ds_shape)
+
+        if pd_name in ["Fx", "Fy", "Fz", "Vx", "Vy", "Vz", "rho", "tags"]:
+            ds_all_primal = np.asarray(patchdatas[pd_name].dataset)
+        elif pd_name == "Bx":
```
[the remark this answers is not in the document]
Comment (rochSmets) [reply pyphare/pyphare/pharesee/run.py]: done !

File: pyphare/pyphare/pharesee/run.py [hunk 26]
```diff
@@ -323,33 +504,92 @@ def GetTags(self, time, merged=False):
...
+            hier.patch_levels,
+            hier.domain_box,
+            refinement_ratio=hier.refinement_ratio,
+            time=time,
+            data_files=hier.data_files,
+        )
+        # return self._get(hier, time, merged, interp)
 
-    def GetNi(self, time, merged=False, interp="nearest"):
+    def GetNi(self, time, merged=False, interp="nearest", all_primal=True):
```
[the remark this answers is not in the document]
Comment (rochSmets) [reply pyphare/pyphare/pharesee/run.py]: that was to get a `ScalarField` and not a `PatchHierarchy`. Now, `GetNi` return  a `ScalarField` wo. needing this kw

File: pyphare/pyphare/core/operators.py (hunk 16 above)
Comment (rochSmets) [reply pyphare/pyphare/core/operators.py:109]: done

File: pyphare/pyphare/core/operators.py (hunk 17 above)
Comment (rochSmets) [reply pyphare/pyphare/core/operators.py:141]: done

File: pyphare/pyphare/core/operators.py (hunk 18 above)
Comment (rochSmets) [reply pyphare/pyphare/core/operators.py:173]: et enplus, ca marche !

File: pyphare/pyphare/core/operators.py (hunk 19 above)
Comment (rochSmets) [reply pyphare/pyphare/core/operators.py:181]: yep, done

**Comment by nicolasaunai** [comment]
Closing because is in #859

**Pull request closed without merging**